### PR TITLE
feat(api-server): add inspect MeshGateway policies endpoint

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

Add new API server inspect policies endpoint. See [discussion for behavior around `Selectors`](https://github.com/kumahq/kuma/issues/3720#issuecomment-1080818092).

Part of #3720

### CODE REVIEW

**Comment by lobkovilya** [comment]
Hmm, do we really need this endpoint? I'm just not sure how useful is it to see what policies were applied for each MeshGateway selector without taking into account real dataplanes?

**Comment by michaelbeaumont** [comment]
> Hmm, do we really need this endpoint? I'm just not sure how useful is it to see what policies were applied for each MeshGateway selector without taking into account real dataplanes?

[I posted my rationale](https://github.com/kumahq/kuma/issues/3720#issuecomment-1083290079) already. I would ask instead, do users actually want to check individual `Dataplane`s? Even for a simple k8s `Deployment`, they are a noisy step in between since they're identical across the `Deployment`.

I think conceptually many users will be applying policies to a `MeshGatewayInstance` or often the `MeshGateway` since I think in a majority of cases there's one `MeshGatewayInstance` (in particular in the Gateway API case).

File: pkg/api-server/inspect_endpoints.go [hunk 1]
```diff
@@ -67,6 +70,80 @@ func getMatchedPolicies(
...
+		dataplane.SetMeta(&rest.ResourceMeta{
+			Type: string(core_mesh.DataplaneType),
+			Name: dataplaneKey.Name,
+			Mesh: dataplaneKey.Mesh,
+		})
+
+		if err := dataplane.SetSpec(&mesh_proto.Dataplane{
+			Networking: &mesh_proto.Dataplane_Networking{
+				Gateway: &mesh_proto.Dataplane_Networking_Gateway{
+					Tags: selector.Match,
```
Comment (lobkovilya) [inline pkg/api-server/inspect_endpoints.go:111]: what happens if the selector has `kuma.io/service: *`?

Comment (michaelbeaumont) [reply pkg/api-server/inspect_endpoints.go:111]: Hmm every Gateway selector must have a service tag in fact. Everything works, as far as I can tell at least

Comment (lobkovilya) [reply pkg/api-server/inspect_endpoints.go:111]: Could you please try the following patch:
```
Index: pkg/api-server/inspect_endpoints_test.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/api-server/inspect_endpoints_test.go b/pkg/api-server/inspect_endpoints_test.go
--- a/pkg/api-server/inspect_endpoints_test.go	(revision 0a268cdaa03c235d993ebcc14d932a1426620935)
+++ b/pkg/api-server/inspect_endpoints_test.go	(date 1648720926821)
@@ -486,7 +486,9 @@
 				&core_mesh.TrafficPermissionResource{
 					Meta: &test_model.ResourceMeta{Name: "tp-1", Mesh: "default"},
 					Spec: &mesh_proto.TrafficPermission{
-						Sources:      anyService(),
+						Sources: selectors{
+							serviceSelector("gateway", ""),
+						},
 						Destinations: anyService(),
 					},
 				},
@@ -494,7 +496,7 @@
 					Meta: &test_model.ResourceMeta{Name: "gateway", Mesh: "default"},
 					Spec: &mesh_proto.MeshGateway{
 						Selectors: selectors{
-							serviceSelector("gateway", ""),
+							serviceSelector("*", ""),
 						},
 						Conf: &mesh_proto.MeshGateway_Conf{
 							Listeners: []*mesh_proto.MeshGateway_Listener{

```
I believe result should be the same, but it's different

Comment (michaelbeaumont) [reply pkg/api-server/inspect_endpoints.go:111]: Ah I see, I didn't see the `: *`, good point!

**Comment by lobkovilya** [comment]
To be honest I'm still not convinced we need this method. As you correctly noticed it's the same as having `kumactl inspect service` which we don't have today. For me, the main problem is that you see that some policy is matched for MeshGateway, but in reality, there could be no dataplanes that actually have this policy. 

In my opinion, debugging starts from the bottom. You noticed that some Dataplane behaves weirdly and you start looking into `config_dump` and check what policies are applied for this specific Dataplane. Also, a lot of users discover things through Kuma GUI where it should be easy to navigate between specific instances.

**Comment by michaelbeaumont** [comment]
> For me, the main problem is that you see that some policy is matched for MeshGateway, but in reality, there could be no dataplanes that actually have this policy.

Yes, while this is possible, it's also a problem with `kumactl inspect service`. I don't think the possibility that users are micromanaging their Dataplanes is an argument against giving them more ergonomic tools based around service tags or gateways.

> In my opinion, debugging starts from the bottom. You noticed that some Dataplane behaves weirdly and you start looking into config_dump and check what policies are applied for this specific Dataplane. Also, a lot of users discover things through Kuma GUI where it should be easy to navigate between specific instances.

Is "this `Dataplane` isn't behaving" the place users start? I mean it's super useful for us developers but I don't know if users want to or should have to look at Envoy config dumps.

Isn't it more likely "my service can't talk to my other service" (`kumactl inspect service`)? Or "requests to my `MeshGateway` at example.com/endpoint" aren't being rate limiting (`kumactl inspect meshgateway`)?

> To be honest I'm still not convinced we need this method. As you correctly noticed it's the same as having kumactl inspect service which we don't have today. 

I don't think that's a reason not to add it, though yes it's part of that wider discussion. I'm convinced it's useful but I guess we can hold until we have a plan for how it fits in with other tools.

**Comment by lobkovilya** [comment]
Ok, you are right, if the code is already there, we can add it and give it a try. But I think we have to fix the problem with `kuma.io/service: *` in selector

**Comment by michaelbeaumont** [comment]
@lobkovilya I actually came around to closing it for now. 😄  It's part of a larger discussion like you suggested, better not to expand the public API until we've thought it through for all resources

**Comment by lobkovilya** [comment]
I agree 😀

**Pull request closed without merging**